app/src/main/java/com/bnyro/contacts/ui/models/ContactsModel.kt

```diff
@@ -49,66 +49,121 @@ class ContactsModel(
             ContactsSource.LOCAL -> localContactsRepository
             ContactsSource.DEVICE -> deviceContactsRepository
         }
-    private val permissions = arrayOf(
-        Manifest.permission.WRITE_CONTACTS,
-        Manifest.permission.READ_CONTACTS
-    )
-    var initialContactId: Long? by mutableStateOf(null)
+    var initialContactId: Long? = null
     var initialContactData: ContactData? by mutableStateOf(null)
 
-    fun loadContacts(context: Context) {
-        isLoading = true
-        if (contactsSource == ContactsSource.DEVICE &&
-            !PermissionHelper.checkPermissions(context, permissions)
-        ) {
-            return
+    var localContacts: ContactListState by mutableStateOf(ContactListState.Loading)
+        private set
+
+    var deviceContacts: ContactListState by mutableStateOf(ContactListState.Loading)
+        private set
+
+    var contacts: List<ContactData>
+        get() = when (contactsSource) {
+            ContactsSource.LOCAL -> (localContacts as? ContactListState.Success)?.contacts.orEmpty()
+            ContactsSource.DEVICE -> (deviceContacts as? ContactListState.Success)?.contacts.orEmpty()
         }
-        viewModelScope.launch(Dispatchers.IO) {
-            isLoading = true
-            try {
-                contacts = contactsRepository.getContactList().toMutableStateList()
-            } catch (e: Exception) {
-                return@launch
+        private set(value) {
+            when (contactsSource) {
+                ContactsSource.LOCAL -> localContacts = ContactListState.Success(value)
+                ContactsSource.DEVICE -> deviceContacts = ContactListState.Success(value)
             }
-            isLoading = false
-            CoroutineScope(Dispatchers.IO).launch {
-                contacts.map {
+        }
+
+    init {
+        loadContacts(context)
+
+        initialContactId?.let {
+            setInitialContactID(it, context)
+        }
+    }
+
+    private suspend inline fun getLocalContacts() {
+        localContacts = try {
+            localContactsRepository.getContactList().takeIf { it.isNotEmpty() }?.let {
+                ContactListState.Success(it)
+            } ?: ContactListState.Empty
+        } catch (_: Exception) {
+            ContactListState.Error
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private suspend inline fun getDeviceContacts(context: Context) {
+        deviceContacts = ContactListState.Loading
+        while (!PermissionHelper.hasPermission(context, Manifest.permission.READ_CONTACTS)) {
```

Comment:
This hurts my feelings 🫠

Reply:
I think we can use `rememberLauncherForActivityResult` inside Compose instead and call the function again upon receiving `ActivityResult.Success`

Reply:
We'll leave that for another day then.

Reply:
IMO, we should show a separate activity/screen and request the permission from the user, and keep that activity  open until the user decide what to do.

If the user grants permission, then we can show the regular app interface.

If the user denies permission, then we can still show the app interface without the **Device Contacts** tab

[Diagram](https://developer.android.com/static/images/training/permissions/workflow-runtime.svg)